crates/sparrow-expressions/src/evaluators.rs

```diff
@@ -0,0 +1,203 @@
+use std::borrow::Cow;
+
+use arrow_array::types::ArrowPrimitiveType;
+use arrow_schema::{DataType, Schema};
+use hashbrown::HashMap;
+use index_vec::IndexVec;
+use itertools::Itertools;
+use sparrow_arrow::scalar_value::ScalarValue;
+use sparrow_physical::Expr;
+
+use crate::evaluator::Evaluator;
+use crate::values::{ArrayRefValue, BooleanValue, PrimitiveValue, StringValue, StructValue};
+use crate::Error;
+
+mod cast;
+mod coalesce;
+mod column;
+mod comparison;
+mod field_ref;
+mod hash;
+mod is_valid;
+mod json_field;
+mod literal;
+mod logical;
+mod macros;
+mod math;
+mod record;
+mod string;
+mod time;
+
+/// Type alias for a function used to create an [Evaluator].
+type EvaluatorFactoryFn =
+    dyn for<'a> Fn(StaticInfo<'a>) -> error_stack::Result<Box<dyn Evaluator>, Error> + Send + Sync;
+
+/// Factory for creating evaluators with a specific name.
+struct EvaluatorFactory {
+    name: &'static str,
+    create: &'static EvaluatorFactoryFn,
+}
+
+inventory::collect!(EvaluatorFactory);
+
+pub struct StaticInfo<'a> {
+    input_schema: &'a Schema,
+    name: &'a Cow<'static, str>,
+    literal_args: &'a [ScalarValue],
+    args: Vec<&'a StaticArg<'a>>,
+    result_type: &'a DataType,
+}
+
+/// Information available when creating evaluators for a query.
+pub struct StaticArg<'a> {
+    /// Expression index of argument.
+    index: usize,
+    /// The DataType of the argument.
+    data_type: &'a DataType,
+}
+
+impl<'a> StaticArg<'a> {
+    pub fn primitive<T: ArrowPrimitiveType>(
+        &self,
+    ) -> error_stack::Result<PrimitiveValue<T>, Error> {
+        PrimitiveValue::try_new(self.index, self.data_type)
+    }
+
+    pub fn boolean(&self) -> error_stack::Result<BooleanValue, Error> {
+        BooleanValue::try_new(self.index, self.data_type)
+    }
+
+    pub fn string(&self) -> error_stack::Result<StringValue, Error> {
+        StringValue::try_new(self.index, self.data_type)
+    }
+
+    pub fn array_ref(&self) -> ArrayRefValue {
+        ArrayRefValue::new(self.index)
+    }
+
+    pub fn struct_(&self) -> error_stack::Result<StructValue, Error> {
```

Comment:
`struct_`?

Reply:
Heh. Yeah. `struct` is a reserved word. There is a weird syntax for using that as a name, but that seems to be frowned upon. You *can* do `r#struct` (which uses a raw identifier and bypasses the reserved word), but based on various things it seems like it is easier / preferred to just use an `_` to make it not a reserved word.

https://www.reddit.com/r/rust/comments/bb3ysz/comment/ekga8in/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

We *could* call this `record`, but that would be better if we called it record everywhere -- we generally use `record` when we talk about Fenl and `struct` when we talk about Arrow (to match `StructArray`) so this felt like the easiest option.